File: src/main/java/ru/superu/rentcarregister/model/Rental.java
package ru.superu.rentcarregister.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;
import ru.superu.rentcarregister.model.enums.RentalStatus;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "rentals")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder(toBuilder = true)
@ToString(exclude = {"driver", "car"})
public class Rental {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "driver_id", nullable = false)
    private Person driver;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "car_id", nullable = false)
    private Car car;

    @NotNull
    private LocalDate plannedStartDate;

    @NotNull
    private LocalDate plannedEndDate;

    private LocalDate actualStartDate;

    private LocalDate actualReturnDate;

    @Min(0)
    private Integer startMileage;

    @Min(0)
    private Integer endMileage;

    @DecimalMin("0.0")
    @DecimalMax("100.0")
    @Column(precision = 5, scale = 2)
    private BigDecimal startFuelPercent;

    @DecimalMin("0.0")
    @DecimalMax("100.0")
    @Column(precision = 5, scale = 2)
    private BigDecimal endFuelPercent;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private RentalStatus status = RentalStatus.PENDING;

    @NotNull
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal dailyRate;

    @DecimalMin("0.0")
    @DecimalMax("100.0")
    @Builder.Default
    @Column(precision = 5, scale = 2)
    private BigDecimal discountPercent = BigDecimal.ZERO;

    @Column(precision = 10, scale = 2)
    private BigDecimal totalAmount;

    @Column(precision = 10, scale = 2)
    private BigDecimal deposit;

    @Column(precision = 10, scale = 2)
    private BigDecimal fuelCompensation;

    @Size(max = 1000)
    @Column(length = 1000)
    private String notes;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @PrePersist
    private void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    private void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public long getPlannedDays() {
        if (plannedStartDate == null || plannedEndDate == null) return 0;
        return java.time.temporal.ChronoUnit.DAYS.between(plannedStartDate, plannedEndDate);
    }

    public long getActualDays() {
        if (actualStartDate == null || actualReturnDate == null) return 0;
        return java.time.temporal.ChronoUnit.DAYS.between(actualStartDate, actualReturnDate);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rental rental = (Rental) o;
        return id != null && id.equals(rental.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: rentcar-flutter/lib/models/rental.dart
import 'package:flutter/material.dart';

enum RentalStatus { pending, active, completed, cancelled, overdue }

extension RentalStatusExt on RentalStatus {
  String get label => switch (this) {
        RentalStatus.pending => 'Ожидает',
        RentalStatus.active => 'Активна',
        RentalStatus.completed => 'Завершена',
        RentalStatus.cancelled => 'Отменена',
        RentalStatus.overdue => 'Просрочена',
      };

  Color get color => switch (this) {
        RentalStatus.pending => Colors.orange,
        RentalStatus.active => Colors.blue,
        RentalStatus.completed => Colors.green,
        RentalStatus.cancelled => Colors.grey,
        RentalStatus.overdue => Colors.red,
      };
}

class Rental {
  final String id;
  final String driverId;
  final String driverName;
  final String carId;
  final String carInfo;
  final String plannedStartDate;
  final String plannedEndDate;
  final String? actualStartDate;
  final String? actualReturnDate;
  final int? startMileage;
  final int? endMileage;
  final double? startFuelPercent;
  final double? endFuelPercent;
  final RentalStatus status;
  final double dailyRate;
  final double discountPercent;
  final double? totalAmount;
  final double? deposit;
  final double? fuelCompensation;
  final int plannedDays;
  final String? notes;
  final String? createdAt;

  const Rental({
    required this.id,
    required this.driverId,
    required this.driverName,
    required this.carId,
    required this.carInfo,
    required this.plannedStartDate,
    required this.plannedEndDate,
    this.actualStartDate,
    this.actualReturnDate,
    this.startMileage,
    this.endMileage,
    this.startFuelPercent,
    this.endFuelPercent,
    required this.status,
    required this.dailyRate,
    required this.discountPercent,
    this.totalAmount,
    this.deposit,
    this.fuelCompensation,
    required this.plannedDays,
    this.notes,
    this.createdAt,
  });

  factory Rental.fromJson(Map<String, dynamic> json) => Rental(
        id: json['id'] as String,
        driverId: json['driverId'] as String,
        driverName: json['driverName'] as String,
        carId: json['carId'] as String,
        carInfo: json['carInfo'] as String,
        plannedStartDate: json['plannedStartDate'] as String,
        plannedEndDate: json['plannedEndDate'] as String,
        actualStartDate: json['actualStartDate'] as String?,
        actualReturnDate: json['actualReturnDate'] as String?,
        startMileage: json['startMileage'] as int?,
        endMileage: json['endMileage'] as int?,
        startFuelPercent: (json['startFuelPercent'] as num?)?.toDouble(),
        endFuelPercent: (json['endFuelPercent'] as num?)?.toDouble(),
        status: RentalStatus.values.firstWhere(
          (e) => e.name.toUpperCase() == json['status'],
          orElse: () => RentalStatus.pending,
        ),
        dailyRate: (json['dailyRate'] as num).toDouble(),
        discountPercent: (json['discountPercent'] as num).toDouble(),
        totalAmount: (json['totalAmount'] as num?)?.toDouble(),
        deposit: (json['deposit'] as num?)?.toDouble(),
        fuelCompensation: (json['fuelCompensation'] as num?)?.toDouble(),
        plannedDays: json['plannedDays'] as int,
        notes: json['notes'] as String?,
        createdAt: json['createdAt'] as String?,
      );
}
